staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -1219,6 +1222,196 @@ func RunSendInitialEventsBackwardCompatibility(ctx context.Context, t *testing.T
 	w.Stop()
 }
 
+func RunWatchSemantics(ctx context.Context, t *testing.T, newStorage func(t *testing.T) (storage.Interface, func(storageRV string), func())) {
```

Comment:
What I'm not a fan of is that every single other test in this package out of 50-ish is taking storage as a parameter.
I would like to figure out a way to do the same here - that is introducing an additional cognitive burden...

One way of achieving this would be to define:
- helper function being a body of the t.Run() here
- a number of dedicated Run* functions that would then be run by imlementations

While far from ideal, I still think I would prefer that.

Reply:
Another, idea might be to try making it a single test with pattern:
- prepare storage [creating init pods]
- create all the the watches and check the init pods
- create "afterEstablish" pods
- check the watches continuation

The only trick with that is that these tests use different "storageResourceVersion", but we need to fix that anyway for etcd.
The way I would do that [and I would actually do that already in place for the existing test before moving it here] would be:
- createPods
- check their RV by doing list with RV=0   [say we're in RV=3]
- do list with RV=<that value> [to ensure that watchcache caught up]
- start watches that you want storage to be in RV=3
- create some auxiliary pod in a different namespace [this will not affect your watches, but will advance your storage RV]
- [as above] wait for watchcache to catch up
- start watches that you want storage to be in RV=4 (you may create two pods above if you want to preserve rv=5)


That will allow you to get rid of storageVersionSetter which we need to do anyway.

Reply:
If you would like to accept `storage.Interface` then what we could do would be to change the test cases to create pods in separate namespaces.

If you want to get rid of the `storageResourceVersion` now (I was planning to do it in the next PR) then what I had in mind was to essentially move the fist test case to a separate test. Please have a look at https://github.com/kubernetes/kubernetes/pull/119557/files#r1316855170

I think this is the only scenario that assumes it can manipulate `storageResourceVersion`.
For etcd implementation it doesn't make sense because initial "LIST" against etcd is strongly consistent.

Reply:
> If you would like to accept storage.Interface then what we could do would be to change the test cases to create pods in separate namespaces.

That sounds reasonable too.

> If you want to get rid of the storageResourceVersion now [...]

Yeah - i would start with it. Separating this testcase makes sense to me.

I would actually still make it a generic one and run against etcd if possible (just for etcd we won't be able to set the RV, but this isn't needed there).

Reply:
> The way I would do that [and I would actually do that already in place for the existing test before moving it here] would be:

> createPods
check their RV by doing list with RV=0 [say we're in RV=3]
do list with RV= [to ensure that watchcache caught up]
start watches that you want storage to be in RV=3
create some auxiliary pod in a different namespace [this will not affect your watches, but will advance your storage RV]
[as above] wait for watchcache to catch up
start watches that you want storage to be in RV=4 (you may create two pods above if you want to preserve rv=5)
That will allow you to get rid of storageVersionSetter which we need to do anyway.


This won't be exactly the same case as the case already defined. 
The main difference will be that the watch-cache will already have been caught up with the db.
For now this is okay because what we we want to test is whether we received the global RV in the bookmark.
We might need to redesign the test if we would like to exercise more edge cases i.e. the watch-cache being behind, or the watch-cache being behind multiple RVs (the first received event should not trigger sending the bookmark).